### DISCUSSION TITLE
spleen segmentation - torch sizes of val_outputs, val_labels

### DISCUSSION THREAD

**Discussion by ctpn** [Project-MONAI/MONAI#1615] [Q&A] [answered]
Hello, 

Apologies if this has already been covered elsewhere as it seems to be a rather basic question.

I am wondering why the torch sizes of val_outputs and val_labels fluctuate; specifically, while the batch size and number of channels remain constant, the length, width, and depth of images seems to fluctuate. If it were constantly decreasing, I would think that this is perhaps due to stride / convolutions, but this does not seem to be the case. 

For context, I am trying my own dataset on the spleen segmentation UNet, but am encountering an indexing error that is not consistently reproducible. In contrast to the spleen data, I noticed that the torch sizes of the val_outputs, val_labels are the same as my images, and thought that this might be part of the issue.

Any advice etc. would be appreciated. Thank you!

```
epoch 16 average loss: 0.5209
val_outputs shape: torch.Size([1, 2, 226, 157, 113])
val_outputs shape: torch.Size([1, 2, 226, 157, 113])
val_labels shape: torch.Size([1, 2, 226, 157, 113])
val_outputs shape: torch.Size([1, 2, 305, 267, 124])
val_outputs shape: torch.Size([1, 2, 305, 267, 124])
val_labels shape: torch.Size([1, 2, 305, 267, 124])
val_outputs shape: torch.Size([1, 2, 289, 262, 108])
val_outputs shape: torch.Size([1, 2, 289, 262, 108])
val_labels shape: torch.Size([1, 2, 289, 262, 108])
val_outputs shape: torch.Size([1, 2, 271, 244, 241])
val_outputs shape: torch.Size([1, 2, 271, 244, 241])
val_labels shape: torch.Size([1, 2, 271, 244, 241])
val_outputs shape: torch.Size([1, 2, 257, 236, 251])
val_outputs shape: torch.Size([1, 2, 257, 236, 251])
val_labels shape: torch.Size([1, 2, 257, 236, 251])
val_outputs shape: torch.Size([1, 2, 230, 165, 198])
val_outputs shape: torch.Size([1, 2, 230, 165, 198])
val_labels shape: torch.Size([1, 2, 230, 165, 198])
val_outputs shape: torch.Size([1, 2, 257, 205, 148])
val_outputs shape: torch.Size([1, 2, 257, 205, 148])
val_labels shape: torch.Size([1, 2, 257, 205, 148])
val_outputs shape: torch.Size([1, 2, 268, 217, 114])
val_outputs shape: torch.Size([1, 2, 268, 217, 114])
val_labels shape: torch.Size([1, 2, 268, 217, 114])
val_outputs shape: torch.Size([1, 2, 233, 209, 101])
val_outputs shape: torch.Size([1, 2, 233, 209, 101])
val_labels shape: torch.Size([1, 2, 233, 209, 101])
current epoch: 16 current mean dice: 0.0924
best mean dice: 0.1066 at epoch: 12
```

**Comment by rijobro** [comment, marked as answer]
Hi, the fluctuations that you are seeing are caused by the transforms applied to the images/labels. In this particular example, it is `CropForegroundd`, which crops the image as small as possible without cropping any non-zero voxels. This depends on the image, and is the reason why you see these differences. You wouldn't see the fluctuations in the training data, however, because the transformation following `CropForegroundd` is `RandCropByPosNegLabeld`, which is creating 4 examples, each of size `(96,96,96)`. 

Having images of different sizes compared to each other is fine, but will cause an error if you have batch size > 1, since the tensors will be concatenated together, which requires them to be the same size. For this reason, you'll notice that in the spleen tutorial, training batch_size > 1 (because they're all `(96,96,96)`), but validation batch_size == 1 (varying sizes).

If you want to increase the batch size for your validation data, then you could pad them all to be the same size, which might be `(305,267,251)` in your example above. Of course, that doesn't generalise very well. I'm in the process of adding two new methods of allowing for data of different sizes:

1. For any given batch, pad all images to be the size of the largest image (in each dimension)
2. Modify the collation function (normally the tensors are concatenated together, but we can modify that if we want).

**Reply by ctpn** [reply]
Thanks very much for your explanation, @rijobro! It helps a lot for debugging and general understanding of transforms, much appreciated!!
